File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Users/ViewControllers/UsersViewController/UserListViewController.m
//
//  UserListViewController.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 30.08.2021.
//  Copyright © 2021 QuickBlox Team. All rights reserved.
//

#import "UserListViewController.h"
#import "UserTableViewCell.h"
#import "Log.h"
#import "QBUUser+Videochat.h"
#import "UIView+Videochat.h"
#import "UIViewController+Alert.h"

@interface UserListViewController ()
@property (nonatomic, assign) Boolean isProcessing;
@end

@implementation UserListViewController
//MARK: - Life Cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    UINib *nibUserCell = [UINib nibWithNibName:kUserCellIdentifier bundle:nil];
    [self.tableView registerNib:nibUserCell forCellReuseIdentifier:kUserCellIdentifier];
    self.tableView.allowsMultipleSelection = YES;
    self.userList = [[UserList alloc] init];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self fetchUsers];
}

//MARK - Setup
- (void)fetchUsers {
    if (self.isProcessing) {
        return;
    }
    self.isProcessing = TRUE;
    [self.refreshControl beginRefreshing];
    __weak __typeof(self)weakSelf = self;
    [self.userList fetchWithPage:1 completion:^(NSArray<QBUUser *> * _Nonnull users, NSError * _Nonnull error) {
        if (users && weakSelf.onFetchedUsers) {
            weakSelf.onFetchedUsers(users);
        }
        [weakSelf.refreshControl endRefreshing];
        [weakSelf.tableView reloadData];
        weakSelf.isProcessing = NO;
    }];
}

- (void)fetchNext {
    __weak __typeof(self)weakSelf = self;
    [self.userList fetchNextWithCompletion:^(NSArray<QBUUser *> * _Nullable users, NSError * _Nullable error) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if (users && strongSelf.onFetchedUsers) {
            strongSelf.onFetchedUsers(users);
        }
        [strongSelf.tableView reloadData];
    }];
}

//MARK: - Actions
- (IBAction)refreshUsers:(UIRefreshControl *)sender {
    [self fetchUsers];
}

//MARK: - Public Methods
- (void)setupSelectedUsers:(NSArray<QBUUser *> *)users {
    [self.userList appendUsers:users];
    for (QBUUser *user in users) {
        [self.userList.selected addObject:@(user.ID)];
    }
}

- (void)removeSelectedUsers {
    for (NSIndexPath *indexPathForSelectedRow in self.tableView.indexPathsForSelectedRows) {
        [self.tableView deselectRowAtIndexPath:indexPathForSelectedRow animated:NO];
    }
    [self.userList.selected removeAllObjects];
}

- (void)removeSelectedUser:(QBUUser *)user {
    NSUInteger index = [self.userList.fetched indexOfObject:user];
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:index inSection:0];
    [self.tableView deselectRowAtIndexPath:indexPath animated:NO];
    [self.userList.selected removeObject:@(user.ID)];
}

#pragma mark - Table view data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.userList.fetched.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UserTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kUserCellIdentifier];
    if (cell == nil) {
        NSArray *topLevelObjects = [[NSBundle mainBundle] loadNibNamed:kUserCellIdentifier owner:self options:nil];
        cell = [topLevelObjects objectAtIndex:0];
    }
    QBUUser *user = self.userList.fetched[indexPath.row];
    [cell setupUserID:user.ID];
    [cell setupUserName:user.name];
    cell.tag = indexPath.row;
    
    NSUInteger lastItemNumber = self.userList.fetched.count - 1;
    if (indexPath.row == lastItemNumber && self.userList.isLoadAll == NO) {
        [self fetchNext];
    }
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (self.userList.selected.count > 2) {
        [tableView deselectRowAtIndexPath:indexPath animated:NO];
    }
    QBUUser *user = self.userList.fetched[indexPath.row];
    [self.userList.selected addObject:@(user.ID)];
    if (self.onSelectUser) {
        self.onSelectUser(user, YES);
    }
}

- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath {
    QBUUser *user = self.userList.fetched[indexPath.row];
    if (![self.userList.selected containsObject:@(user.ID)]) {
        return;
    }
    [self.userList.selected removeObject:@(user.ID)];
    if (self.onSelectUser) {
        self.onSelectUser(user, NO);
    }
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(nonnull UITableViewCell *)cell forRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    QBUUser *user = self.userList.fetched[indexPath.row];
    if ([self.userList.selected containsObject:@(user.ID)]) {
        [self.tableView selectRowAtIndexPath:indexPath animated:NO scrollPosition:UITableViewScrollPositionNone];
    } else {
        [self.tableView deselectRowAtIndexPath:indexPath animated:NO];
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Users/ViewControllers/UserListViewController.swift
//
//  UserListViewController.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 28.12.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox

typealias SelectUserCompletion = ( _ user: QBUUser, _ isSelected: Bool) -> Void
typealias FetchedUsersCompletion = ( _ users: [QBUUser]) -> Void

class UserListViewController: UITableViewController {
    //MARK: - Properties
    var userList = UserList()
    var onSelectUser: SelectUserCompletion?
    var onFetchedUsers: FetchedUsersCompletion?
    private var isProcessing = false

    //MARK: - Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.register(UINib(nibName: UserCellConstant.reuseIdentifier, bundle: nil),
                           forCellReuseIdentifier: UserCellConstant.reuseIdentifier)
        tableView.allowsMultipleSelection = true
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        fetchUsers()
    }
    
    //MARK - Setup
    func fetchUsers() {
        if isProcessing == true {
            return
        }
        isProcessing = true
        refreshControl?.beginRefreshing()
        userList.fetchWithPage(1) { [weak self] (users, error) in
            guard let self = self, let users = users else {
                self?.refreshControl?.endRefreshing()
                self?.isProcessing = false
                return
            }
            if users.isEmpty == false {
                self.onFetchedUsers?(users)
            }
            self.refreshControl?.endRefreshing()
            self.tableView.reloadData()
            self.isProcessing = false
        }
    }
    
    func fetchNext() {
        userList.fetchNext { [weak self] (users, error) in
            guard let self = self, let users = users else {
                return
            }
            if users.isEmpty == false {
                self.onFetchedUsers?(users)
            }
            self.tableView.reloadData()
        }
    }

    //MARK: - Actions
    @IBAction func refreshUsers(_ sender: UIRefreshControl) {
        fetchUsers()
    }
    
    //MARK: - Public Methods
    func setupSelectedUsers(_ users: [QBUUser]) {
        userList.append(users)
        for user in users {
            userList.selected.insert(user.id)
        }
    }
    
    func removeSelectedUsers() {
        guard let indexPathsForSelectedRows = tableView.indexPathsForSelectedRows else {
            return
        }
        for indexPathForSelectedRow in indexPathsForSelectedRows {
            tableView.deselectRow(at: indexPathForSelectedRow, animated: false)
        }
        userList.selected.removeAll()
    }
    
    func removeSelectedUser(_ user: QBUUser) {
        if let index = userList.fetched.firstIndex(of: user)  {
            let indexPath = IndexPath(row: index, section: 0)
            tableView.deselectRow(at: indexPath, animated: false)
        }
        userList.selected.remove(user.id)
    }
    
    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return userList.fetched.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: UserCellConstant.reuseIdentifier,
                                                       for: indexPath) as? UserTableViewCell else {
            return UITableViewCell()
        }
        let user = userList.fetched[indexPath.row]
        cell.userColor = user.id.generateColor()
        cell.userNameLabel.text = user.fullName ?? user.login
        cell.userAvatarLabel.text = String(user.fullName?.capitalized.first ?? Character("U"))
        cell.tag = indexPath.row
        
        let lastItemNumber = userList.fetched.count - 1
        if indexPath.row == lastItemNumber, userList.isLoadAll == false {
            fetchNext()
        }
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let user = userList.fetched[indexPath.row]
        if userList.selected.count > 2 {
            tableView.deselectRow(at: indexPath, animated: false)
            onSelectUser?(user, true)
            return
        }
        userList.selected.insert(user.id)
        onSelectUser?(user, true)
    }
    
    override func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
        let user = userList.fetched[indexPath.row]
        if userList.selected.contains(user.id) == false {
            return
        }
        userList.selected.remove(user.id)
        onSelectUser?(user, false)
    }
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        let user = self.userList.fetched[indexPath.row]
        if userList.selected.contains(user.id) {
            tableView.selectRow(at: indexPath, animated: false, scrollPosition: .none)
            return
        }
        tableView.deselectRow(at: indexPath, animated: false)
    }
}
